src/realm/object-store/sectioned_results.cpp

```diff
@@ -0,0 +1,515 @@
+    }
+    else {
+        m_key = key;
+    }
+}
+
+bool ResultsSection::is_valid() const
+{
+    util::CheckedUniqueLock lock(m_parent->m_mutex);
+    if (!m_parent->is_valid())
+        return false;
+    // See if we need to recalculate the sections before
+    // searching for the key.
+    m_parent->calculate_sections_if_required();
+    if (m_parent->m_sections.find(m_key) == m_parent->m_sections.end())
+        return false;
+
+    return true;
+}
+
+Mixed ResultsSection::operator[](size_t idx) const
+{
+    util::CheckedUniqueLock lock(m_parent->m_mutex);
+    if (!m_parent->is_valid())
+        throw std::logic_error("The parent Results collection is not valid.");
+    m_parent->calculate_sections_if_required();
+    auto it = m_parent->m_sections.find(m_key);
+    REALM_ASSERT(it != m_parent->m_sections.end());
+    auto s = it->second.indices.size();
+    if (idx >= s)
+        std::out_of_range(util::format("Requested index %1 greater than max %2", idx, s - 1));
+    return m_parent->m_results.get_any(it->second.indices[idx]);
+}
+
+Mixed ResultsSection::key()
+{
+    if (!is_valid())
+        throw std::logic_error("This ResultsSection is not in a valid state.");
+    return m_key;
+}
+
+size_t ResultsSection::index()
+{
+    util::CheckedUniqueLock lock(m_parent->m_mutex);
+    if (!m_parent->is_valid())
+        throw std::logic_error("The parent Results collection is not valid.");
+    m_parent->calculate_sections_if_required();
+    auto it = m_parent->m_sections.find(m_key);
+    REALM_ASSERT(it != m_parent->m_sections.end());
+    return it->second.index;
+}
+
+size_t ResultsSection::size()
+{
+    util::CheckedUniqueLock lock(m_parent->m_mutex);
+    if (!m_parent->is_valid())
+        throw std::logic_error("The parent Results collection is not valid.");
+    m_parent->calculate_sections_if_required();
+    auto it = m_parent->m_sections.find(m_key);
+    REALM_ASSERT(it != m_parent->m_sections.end());
+    return it->second.indices.size();
+}
+
+NotificationToken ResultsSection::add_notification_callback(SectionedResultsNotificatonCallback callback,
+                                                            KeyPathArray key_path_array) &
+{
+    return m_parent->add_notification_callback_for_section(m_key, std::move(callback), key_path_array);
+}
+
+SectionedResults::SectionedResults(Results results, SectionKeyFunc section_key_func)
+    : m_results(results)
+    , m_callback(std::move(section_key_func))
+{
+}
+
+SectionedResults::SectionedResults(Results results, Results::SectionedResultsOperator op,
+                                   util::Optional<StringData> prop_name)
+    : m_results(results)
+    , m_callback(builtin_comparison(results, op, prop_name))
+{
+}
+
+void SectionedResults::calculate_sections_if_required()
+{
+    if (m_results.m_update_policy == Results::UpdatePolicy::Never)
+        return;
+    else if ((!m_results.has_changed() || m_results.is_frozen()) && has_performed_initial_evalutation)
```

Comment:
`has_changed()` does a lot more work than `is_frozen()`, so it' d be better to check `is_frozen()` first.